MDEV-31925 mysqld_upgrade --check-if-upgrade-is-needed leaks files

hsser opened the pull request:
<!--
Thank you for contributing to the MariaDB Server repository!

You can help us review your changes faster by filling in this template <3

If you have any questions related to MariaDB or you just want to hang out and meet other community members, please join us on https://mariadb.zulipchat.com/ .
-->

<!--
If you've already identified a https://jira.mariadb.org/ issue that seems to track this bug/feature, please add its number below.
-->
- [x] *The Jira issue number for this PR is: [MDEV-31925*](https://jira.mariadb.org/browse/MDEV-31925)

<!--
An amazing description should answer some questions like:
1. What problem is the patch trying to solve?
2. If some output changed that is not visible in a test case, what was it looking like before the change and how it's looking with this patch applied?
3. Do you think this patch might introduce side-effects in other parts of the server?
-->
## Description
This patch addresses a file leak issue identified in `mysql_upgrade` with the `--check-if-upgrade-is-needed` option. Previously, the execution of `mysql_upgrade` with this flag would leave temporary files in the system's temp directory. This patch ensures that any temporary files created during the `mysql_upgrade `process are properly cleaned up, thus maintaining a cleaner system environment.

## How can this PR be tested?

The patch includes an updated test case that verifies the absence of temporary files after running `mysql_upgrade` with the `--check-if-upgrade-is-needed` option. This test should pass without any temporary files being left behind in the temp directory.
<!--
In many cases, this will be as simple as modifying one `.test` and one `.result` file in the `mysql-test/` subdirectory.
Without automated tests, future regressions in the expected behavior can't be automatically detected and verified.
-->

<!--
Tick one of the following boxes [x] to help us understand if the base branch for the PR is correct. (Currently the earliest maintained branch is 10.3)
-->
## Basing the PR against the correct MariaDB version
- [ ] *This is a new feature and the PR is based against the latest MariaDB development branch.*
- [x] *This is a bug fix and the PR is based against the earliest maintained branch in which the bug can be reproduced.*

<!--
  All code merged into the MariaDB codebase must meet a quality standard and codying style.
  Maintainers are happy to point out inconsistencies but in order to speed up the review and merge process we ask you to check the CODING standards.
-->
## PR quality check
- [x] I checked the [CODING_STANDARDS.md](https://github.com/MariaDB/server/blob/11.0/CODING_STANDARDS.md) file and my PR conforms to this where appropriate.
- [x] For any trivial modifications to the PR, I am ok with the reviewer making the changes themselves.

grooverdan commented on mysql-test/t/mysql_upgrade_file_leak.test:
```diff
@@ -0,0 +1,42 @@
+        unlink("$tmpdir/$file") or warn "Could not remove temporary file $file: $!";
+    }
+}
+die "Test failed: Found temporary file left by mysql_upgrade\n" if $temp_file_found;
```
my perl isn't the best but could this be just `die ... if (!glob "mysql_upgrade*");`?

grooverdan commented on mysql-test/t/mysql_upgrade_file_leak.test:
```diff
@@ -0,0 +1,42 @@
+--echo Perl script completed
+let $MYSQLD_DATADIR= `select @@datadir`;
+--remove_file $MYSQLD_DATADIR/mysql_upgrade_info
+--echo End of test 
```
`--echo End of 10.4 tests` for new convention.

grooverdan commented on mysql-test/t/mysql_upgrade_file_leak.test:
```diff
@@ -0,0 +1,42 @@
+-- source include/mysql_upgrade_preparation.inc
+-- source include/have_working_dns.inc
+-- source include/have_innodb.inc
+-- source include/have_partition.inc
```
can omit these except the first one right?

grooverdan commented on mysql-test/t/mysql_upgrade_file_leak.test:
```diff
@@ -0,0 +1,42 @@
+-- source include/have_innodb.inc
+-- source include/have_partition.inc
+
+set sql_mode="";
```
unused?

grooverdan reviewed (changes requested):
Thanks for the contribution. In addition to the above comments, can you rebase this on 10.4, our earliest maintained version.

Can you squash the 3 commits together.
Also with the commit message can you describe the high level problem and fix.

an3l commented:
Hi,
just to help you with Git to update to 10.4.
It should go like this:
```bash
# Create 10.4 if not exists (where upstream is name of remote upstream or origin `git remote -v`)
$ git branch 10.4 upstream/10.4
# Work on new branch 
$ git checkout 10.4
$ git checkout -b <my-10.4-branch>
# Cherry-pick what you have done
$ git cherry-pick <commit-hash>
# Push changes without closing PR
$ git push <my-10.4-branch>:10.2-31925 -f
```
Change the base in GitHub from 10.2 to 10.4.

hsser commented:
@an3l Thank you so much for helping me out with the Git rebase mess. Seriously, I was pretty lost there, and your guidance was a lifesaver.

hsser commented:
> Thanks for the contribution. In addition to the above comments, can you rebase this on 10.4, our earliest maintained version.
> 
> Can you squash the 3 commits together. Also with the commit message can you describe the high level problem and fix.

@grooverdan Thank you for your comprehensive feedback on the test case. Based on your suggestions, I have made the necessary updates and amendments. Additionally, I have rebased the branch and squashed the commits to maintain a clean history.

grooverdan commented:
Thank you, sorry one last request - can you put the MDEV number at the top of the commit message in the title and then git push --force it to update this PR again.

Looks very clean. Thanks @an3l for helping witg rebase instructions,

grooverdan reviewed (approved):
Thank you.
